pkg/events/manager.go

```diff
@@ -0,0 +1,114 @@
+package events
+
+import (
+	"context"
+	"sync"
+
+	"github.com/filecoin-project/lassie/pkg/types"
+)
+
+// EventManager is responsible for dispatching events to registered subscribers.
+// Events are dispatched asynchronously, so subscribers should not assume that
+// events are received within the window of a blocking retriever.Retrieve()
+// call.
+type EventManager struct {
+	ctx         context.Context
+	lk          sync.RWMutex
+	idx         int
+	started     bool
+	subscribers map[int]types.RetrievalEventSubscriber
+	events      chan types.RetrievalEvent
+	cancel      context.CancelFunc
+	stopped     chan struct{}
+}
+
+// NewEventManager creates a new EventManager. Start() must be called to start
+// the event loop.
+func NewEventManager(ctx context.Context) *EventManager {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	return &EventManager{
+		ctx:         ctx,
+		cancel:      cancel,
+		subscribers: make(map[int]types.RetrievalEventSubscriber),
+		events:      make(chan types.RetrievalEvent, 16),
```

Comment:
Was there rationale around 16 as the buffer size here? Just something big enough to not block?

Reply:
That is my guess. In the context where you're handing off to arbitrary user defined subscribers that could be time consuming, even this my not be enough to gaurantee not blocking.

@rvagg have you looked at the code that ultimately powers the subscriber interfaces on data transfer -- it might be worth borrowing from -- we originally had things setup with just a buffer but found cases where we could exceed the buffer.

https://github.com/filecoin-project/go-statemachine/blob/master/fsm/fsm.go#L159

Nowadays I'd probably use a linked list for this since I worry golang slices would grow arbitrarily here.

Reply:
ultimately this may be premature observation but just wanted to make you all aware of the concern

Reply:
No no, I had the same concern. My thought was if this is not enough, we'd need a data structure to dump the events into. A linked list wouldn't be awful.